src/lib.rs

```diff
@@ -353,6 +353,68 @@ fn mul_and_decimalize(a: u128, b: u128) -> u128 {
 
     // Divide
 
+    // Or just multiply by: ceil(2^321/10^20) then right shift 321 times.
+    // 42719740718418201647900434123391042292054090447133055398940832156444394515613
+
+    if true {
+        use ethnum::U256;
+        // This needs to do multiplication in a way that expands,
+        // since we throw away 321 bits we care only about the higher end, not lower.
+        // So like need to do high low mult with 2 U256's and then bitshift.
+        // I bet this has a lot of room for multiplication optimization.
```

Comment:
For sure! Here's the `U256` multiply function, for reference: https://github.com/nlordell/ethnum-rs/blob/97e71c5d128c2614fb2189f20250b564183c1b94/src/intrinsics/native/mul.rs#L14

I bet copy/pasting that function in and marking it as `#[inline(always)]` would help right away, plus we could pass it 4 arguments (`hi1`, `lo1`, `hi2`, `lo2`) instead of `U256`s, and have it return plain `{ hi: u128, lo: u128 }` structs to make sure it's using registers as much as possible.

We could also get rid of a ton of pointer dereferences - those are an API choice the library makes, but they definitely aren't necessary here.